### ISSUE TITLE
Gender&Age pretrained model

### ISSUE THREAD

**Issue by nttstar** [deepinsight/insightface#272]
Link: [here](https://pan.baidu.com/s/1f8RyNuQd7hl2ItlV-ibBNQ)

How to use:  refer to `get_ga()` method [here](https://github.com/deepinsight/insightface/blob/master/deploy/face_model.py#L95)

Any usage, performance and accuracy issue can be posted here.

**Comment by goodtogood** [comment]
would you like give us some details about this GA model ?

**Comment by alphinside** [comment]
Hi, can you upload the weight to a google drive? Thank you

**Comment by glorioushedgehog** [comment]
I tested gender recognition on [gender-color-feret](http://mivia.unisa.it/datasets/video-analysis-datasets/gender-recognition-dataset/) and got 89.71%.

Interestingly, all the incorrect matches were saying that a woman was a man.

**Comment by nttstar** [comment]
@glorioushedgehog It was trained on Asian dataset.

**Comment by glorioushedgehog** [comment]
I see. Are you planning on training with another dataset too?

**Comment by glorioushedgehog** [comment]
About how many images were used to train this model? I could train one on about 1000 images of different races, but I don't know if that would work well at all.

**Comment by gwsbhqt** [comment]
@nttstar thank you very much, the unexpected surprise!

**Comment by hfuCN** [comment]
Does anyone use this model with flask?When I host the model on flask, the first post request works well, but the second get segfault. Has anyone met?

**Comment by gwsbhqt** [comment]
@hfuCN I'm sorry I'm using tornado, and I don't have time to test this at the moment. I suggest you check if you have loaded the model many times, and you can consider a single mode.
其实我怀疑这里都是中国人(((φ(◎ロ◎;)φ)))

**Comment by ronyuzhang** [comment]
请教下， 你的性别年龄模型是根据那篇论文复现的。

**Comment by ronyuzhang** [comment]
能不能公开年龄性别数据集 或者rec格式

**Comment by ahkarami** [comment]
Dear @gwsbhqt
Thank you for your time & response.

**Comment by minhaj3** [comment]
Hi @nttstar , Can you please also provide the training script for this gamodel(gender age model)? I want to train this model on some another dataset but I dont think train_softmax.py will work for this. Any hints anyone?
Thanks in advance.

**Comment by wayen820** [comment]
I have a mxnet version implementation of SSR-Net for age and gender Estimation,train on megaage-asia,model size just 170kb. [https://github.com/wayen820/gender_age_estimation_mxnet](url)

**Comment by gwsbhqt** [comment]
@wayen820 Really? Is there any public report on the accuracy test of age and gender? Look forward to the performance of your model.

**Comment by boyang24** [comment]
@nttstar here has a problem,in the /deploy/test.py,i use the gamodel you applied the gamodel-r50,tom hanks.png,but the result is age:12,i think this result is not that good,is this normol after use gamodel-r50?

**Comment by nttstar** [comment]
My age result on tom_hanks is 30+. Make sure your pre-processing is correct.

**Comment by wayen820** [comment]
@gwsbhqt  You can refer to the Author's paper, our implementation has achieved better results

**Comment by boyang24** [comment]
@nttstar 
parser.add_argument('--model', default='/home/tuxiang/yangbo/insightface-master/model/model,15', help='path to load model.')
parser.add_argument('--ga-model', default='/home/tuxiang/yangbo/insightface-master/deploy/gamodel-r50/model,0', help='path to load model.')
i download this pre-model,and nothoing changed next,the age output is :[-0.00887062 -0.04818811  0.01538625 -0.02213596  0.01085652 -0.01821288
 -0.01626354  0.02557316  0.0018869  -0.05840746]
1
12
[ 0.00181901 -0.02371245 -0.027376    0.01064258 -0.04279873 -0.12093599
 -0.01720401 -0.05790558 -0.0150299   0.0357873 ]
1.86323
0.0683836
please give me the guide for this problem

**Comment by nttstar** [comment]
@wayen820 It looks nice. Thank you.

**Comment by nttstar** [comment]
@boyang24 Did you align the face?

**Comment by boyang24** [comment]
@nttstar yes  the entire code is parser.add_argument('--image-size', default='112,112', help='')
parser.add_argument('--model', default='/home/tuxiang/yangbo/insightface-master/model/model,15', help='path to load model.')
parser.add_argument('--ga-model', default='/home/tuxiang/yangbo/insightface-master/deploy/gamodel-r50/model,0', help='path to load model.')
parser.add_argument('--gpu', default=0, type=int, help='gpu id')
parser.add_argument('--det', default=0, type=int, help='mtcnn option, 1 means using R+O, 0 means detect from begining')
parser.add_argument('--flip', default=0, type=int, help='whether do lr flip aug')
parser.add_argument('--threshold', default=1.24, type=float, help='ver dist threshold')
args = parser.parse_args()

model = face_model.FaceModel(args)
img = cv2.imread('Tom_Hanks_54745.png')
img = model.get_input(img)
f1 = model.get_feature(img)
print(f1[0:10])
gender, age = model.get_ga(img)
print(gender)
print(age)
the age result is :12
thanks again

**Comment by nttstar** [comment]
Here is my output, age is 35.
========

[deepinsight@ctum2e1402005 deploy]$ python test.py --ga-model ../gamodel-r50/model,0
loading ../gamodel-r50/model 0
[13:05:13] src/nnvm/legacy_json_util.cc:209: Loading symbol saved by previous version v1.2.0. Attempting to upgrade...
[13:05:13] src/nnvm/legacy_json_util.cc:217: Symbol successfully upgraded!
[13:05:16] src/nnvm/legacy_json_util.cc:209: Loading symbol saved by previous version v0.8.0. Attempting to upgrade...
[13:05:16] src/nnvm/legacy_json_util.cc:217: Symbol successfully upgraded!
[13:05:16] src/nnvm/legacy_json_util.cc:209: Loading symbol saved by previous version v0.8.0. Attempting to upgrade...
[13:05:16] src/nnvm/legacy_json_util.cc:217: Symbol successfully upgraded!
[13:05:16] src/nnvm/legacy_json_util.cc:209: Loading symbol saved by previous version v0.8.0. Attempting to upgrade...
[13:05:16] src/nnvm/legacy_json_util.cc:217: Symbol successfully upgraded!
[13:05:16] src/nnvm/legacy_json_util.cc:209: Loading symbol saved by previous version v0.8.0. Attempting to upgrade...
[13:05:16] src/nnvm/legacy_json_util.cc:217: Symbol successfully upgraded!
[13:05:17] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:109: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
1
35

**Comment by boyang24** [comment]
@nttstar i got something wired，firstly，i test this project under python3.5，but i found some trouble is hard to solve，then i switched to python2.7，under python2.7，the age is 12，but for python3.5，the age is 35，the output in 2.7 is：
/usr/local/lib/python2.7/dist-packages/sklearn/utils/fixes.py:313: FutureWarning: numpy not_equal will not check object identity in the future. The comparison did not return the same result as suggested by the identity (`is`)) and will change.
  _nan_object_mask = _nan_object_array != _nan_object_array


but in 3.5 no this kinda problem！i do not know the reason
thanks a lot

**Comment by mayidu** [comment]
I use SSR-Net in gender and age, and the result of age is good, but the gender is bad. And often female to male, even I retrained with my data, the result not improve. So I want to know, how the profrom of mxnet-SSR-Net and insightface gender and age ? @nttstar @wayen820 
Thank you!

**Comment by mayidu** [comment]
Can you tell me what training set is used to train gender age model with you, and what is the training method? If I want to use my own data training, how can I do? Thank you!

**Comment by wuyuanyuan1990** [comment]
> @glorioushedgehog It was trained on Asian dataset.

is that the MegaAge-Asian dataset ? thank you

**Comment by Simplesss** [comment]
Hello, when I use the model of the pre - trained in link through the file model- slim. py .Conversion, convert the parameter file is 235M , and I will be with you within project has converted documents do contrast, found not same with network architecture, so I want to ask you, is my translation has a problem, or this .py is not the file conversion?

**Comment by nttstar** [comment]
Checkout the new GA model under `https://github.com/deepinsight/insightface/tree/master/gender-age`

**Comment by Simplesss** [comment]
> Checkout the new GA model under `https://github.com/deepinsight/insightface/tree/master/gender-age`

First of all, thank you for your reply. But when I used your gender-age model to convert it into NCNN format, I found that the 202 dimensional vectors generated were almost the same, which resulted in 0 and 37 gender and age output.Do you know what the problem is?

**Comment by stereomatchingkiss** [comment]
> paper

Which paper?What is the name?Could you show us the link?
Another model with 1MB, which paper it based on?

**Comment by MohanKrishna-RC** [comment]
T

> Checkout the new GA model under `https://github.com/deepinsight/insightface/tree/master/gender-age`

For the age-gender model you provided, how to prepare the .rec files for training?

**Comment by nobody-cheng** [comment]
你好，请问训练年龄性别数据集的rec的组成格式是怎样的格式，谢谢

**Comment by boyang24** [comment]
训练的数据是公开数据集，格式就是那个自带格式，我并没有重新生成数据





在 2019-05-31 10:59:12，"nobody-cheng" <notifications@github.com> 写道：


你好，请问训练年龄性别数据集的rec的组成格式是怎样的格式，谢谢

**Comment by nobody-cheng** [comment]
您好， 
我将gender-age预测模型转为tf格式后进行预测，发现效果差很多，想知道是不是mtcnn加的LNet网络使得输出更加精细（gender-age工程中的mtcnn对齐我看比官网的加多了一个LNet网络）
谢谢

**Comment by SongQingChe** [comment]
> 训练的数据是公开数据集，格式就是那个自带格式，我并没有重新生成数据 在 2019-05-31 10:59:12，"nobody-cheng" <notifications@github.com> 写道： 你好，请问训练年龄性别数据集的rec的组成格式是怎样的格式，谢谢 — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or mute the thread.

麻烦问一下是什么数据集呀~万分感谢!

**Comment by xysong1201** [comment]
@nttstar Hi, I have 2 questions concerning your age-gender model
1. Could you explain why the final output for the age estimation, there are 200 classes?
How you calculate the label for age? I saw you use np.argmax(), could you detail a little bit more?
2. And what is the loss function you used for gender-age? I found 3 evaluation metric, one is accuracy, one is MAE and I'm confused about the third one?

Thank you in advance

**Comment by ucasiggcas** [comment]
有没有简单的测试脚本啊？多谢大佬

**Comment by yunchangxiaoguan** [comment]
> Does anyone use this model with flask?When I host the model on flask, the first post request works well, but the second get segfault. Has anyone met?

兄弟，你的这个flask解决了嘛，我也想用flask写出接口来，总是不成功

**Comment by bruceche11** [comment]
百度网盘链接失效了  这里奉上新的
链接: https://pan.baidu.com/s/10oHvGo4yE6vi75PCBi6rYw 提取码: u353

**Comment by hailcpy** [comment]
@nttstar the link to gender-age model is not working. I can't download the model.

**Comment by ChrisKong93** [comment]
@Simplesss i have same problem，i convert it into onnx format,  which resulted in 0 and 37 gender and age output.did you solve it?

**Comment by LeeREVO** [comment]
Not able to download from this link anymore, could someone provide a new link?

**Comment by nhatduy19599** [comment]
> Link: [here](https://pan.baidu.com/s/1f8RyNuQd7hl2ItlV-ibBNQ)
> 
> How to use: refer to `get_ga()` method [here](https://github.com/deepinsight/insightface/blob/master/deploy/face_model.py#L95)
> 
> Any usage, performance and accuracy issue can be posted here.

the link of 'gamodel-r50' is invalid, is there any way to download the file 'gamodel-r50' ?
tks.

**Comment by nttstar** [comment]
This model is out of date. Please use the latest insightface package and it contains a lightweight gender&age attribute model.

**Comment by MyraBaba** [comment]
@nttstar 

Hi is there any way to convert old mxnet based ga model to onnx ? According to our test the old mxnet based one is more accurate the current ones. Do you have any chance to compare them ? 

We couldnt convert the old ga model to onnx bytheway
